### ISSUE TITLE
docs/ instructions

### ISSUE THREAD

**Issue by ispysoftware** [Microsoft/WebRTC-universal-samples#45]
I'm adding code to the WrapperGlue project and want to expose it via the Org.WebRtc dll but i'm struggling to figure out how to do this. All the files under the wrapper directory have

//Generated by zsLibEventingTool 

.. at the top but I can't find any instructions on how to run this to generate the files in the wrapper - rebuilding the project (which is what it says to do in the readme.txt file) isn't generating new wrapper code. Are there any docs for this? The docs in the solution folder are just empty templates.

**Comment by ispysoftware** [comment]
Yeah any pointers would be greatly appreciated - we've hit a bit of a wall here

**Comment by robin-raymond** [comment]
@SergioOpenPeer As part of the python scripting can you ensure the regeneration is documented properly and I'll work on documenting more how the IDL works (including the binding process).

**Comment by SergioOpenPeer** [comment]
@ispysoftware If you want to regenerate files generated from idl, you have to delete .flg files from this folder webrtc-uwp-sdk\webrtc\xplatform\webrtc\third_party\idl\zsLib-eventing , and to be sure that everything is regenerate properly you can delete generated files that are in this folder webrtc-uwp-sdk\webrtc\xplatform\webrtc\sdk\windows\wrapper\generated

**Comment by SergioOpenPeer** [comment]
In couple of days, new scripts for environment preparation and building will be published, and there will be option for cleaning idl generated files and their flags.

**Comment by SergioOpenPeer** [comment]
@ispysoftware Just run build. It will be regenerated during build process

**Comment by ispysoftware** [comment]
Oh OK - so it regenerated the bindings for webrtc but not for the code that i've added into the WrapperGlue project. I've added a class and a header in to do custom video capture which is building but I can't access it via the org.webrtc library. What do I need to do to be able to call this code?

**Comment by robin-raymond** [comment]
@SergioOpenPeer Can you give a "hack" they can use right now to force regeneration? I know deleting *.flg files and deleting webrtc/out will do it but there might be a more optimal way.

**Comment by ispysoftware** [comment]
Maybe we're doing something wrong here.. We just want to add a new way of sending frames to webrtc so what we've done is create a new  .cpp and .h file in the wrapperglue project to do that, under namespace webrtc and now we just need to be able to call that via the org.webrtc dll - is that going to work or do we have the approach all wrong? What is it we need to do to get this working? 

Thanks for your responses by the way, it's much appreciated.

**Comment by robin-raymond** [comment]
@ispysoftware You do need to make the IDL regenerate to make it work. IDL does not regenerate all the time and there's an ".flg" file that prevents the regeneration (intentionally) but there's also something inside webrtc ninja that prevents the regeneration as well (not sure where that part is located).

**Comment by jamescadd** [comment]
Maybe I'm missing something, but have you added the new APIs to the IDL before regenerating?

**Comment by ispysoftware** [comment]
Maybe this would just be easier as a feature request - as I feel like this rabbit hole is going to go pretty deep. We need to be able to setup custom audio and video sources - so not using devices, just push audio and video frames into the renderer as byte pointers - like:

```
void VideoCapturer::PushFrame(uint8_t* videoFrame, int videoFrameLength) {
      int stride_y = width_;
      int stride_uv = (width_ + 1) / 2;
      int target_width = width_;
      int target_height = height_;

	  rtc::scoped_refptr<I420Buffer> buffer = I420Buffer::Create(target_width, abs(target_height), stride_y, stride_uv, stride_uv);

	  libyuv::RotationMode rotation_mode = libyuv::kRotate0;

	  const int conversionResult = libyuv::ConvertToI420(
          videoFrame, videoFrameLength, buffer.get()->MutableDataY(),
          buffer.get()->StrideY(), buffer.get()->MutableDataU(),
          buffer.get()->StrideU(), buffer.get()->MutableDataV(),
          buffer.get()->StrideV(), 0, 0,  // No Cropping
              width_, height_, target_width, target_height, rotation_mode,
          FOURCC_24BG);
      if (conversionResult < 0) {
        RTC_LOG(LS_ERROR) << "Failed to convert capture frame from type "
                          << static_cast<int>(FOURCC_24BG) << " to I420.";
        return;
      }
      int64_t captureTime = 0;
      VideoFrame captureFrame(buffer, 0, rtc::TimeMillis(), kVideoRotation_0);
      captureFrame.set_ntp_time_ms(captureTime);
      OnFrame(captureFrame, captureFrame.width(), captureFrame.height());
	}
```

**Comment by jamescadd** [comment]
We need this internally as well. Could you email the request and I'll add the team working on it to discuss? But yes it's a good amount of work. Here's more info on IDL: https://github.com/webrtc-uwp/webrtc-uwp-sdk/wiki/WebRTC-UWP-and-IDL

**Comment by ispysoftware** [comment]
@jamescadd sent you an email, please let me know if you get it

**Comment by jamescadd** [comment]
@ispysoftware email received/replied.  Keeping this open until the documentation is added to the above and we can go from there.
